File: week2/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        HashSet<String> set = new HashSet<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                char n = board[i][j];
                if (n != '.' 
                    && (!set.add("r" + String.valueOf(i) + n)
                    || !set.add("c" + String.valueOf(j) + n)
                    || !set.add("b" + String.valueOf(i/3) + String.valueOf(j/3) + n))) {
                    return false;
                }
            }
        }
        return true;    
    }
}

File: 2026/week2/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        quadrants = [set() for i in range(9)]
        rows = [set() for i in range(9)]
        cols = [set() for i in range(9)]

        for i, row in enumerate(board):
            for j, current in enumerate(row):
                if current == ".":
                    continue
                if not self.addOrInvalidate(rows[i], current):
                    return False
                if not self.addOrInvalidate(cols[j], current):
                    return False
                if not self.addOrInvalidate(quadrants[(i // 3) * 3 + (j // 3)], current):
                    return False
        return True
    
    def addOrInvalidate(self, s: set, v: int) -> bool:
        if v in s:
            return False
        s.add(v)
        return True
